### PR TITLE
listener: allow overriding static listeners via LDS

### LINKED ISSUE #39510: Question - How to initialize envoy with a listener which is overridden by the control plane after
How can a listener be setup in the bootstrap envoy config which can later be overridden through the control plane?

### PR DESCRIPTION
**Pull request by AntonKanug**
Fixes https://github.com/envoyproxy/envoy/issues/39510

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/config/listener/v3/listener.proto [hunk 1]
```diff
@@ -447,4 +447,16 @@ message Listener {
...
   core.v3.TcpKeepalive tcp_keepalive = 37;
+
+  // If true, this static listener may be replaced by an LDS listener with the same
+  // :ref:`name <envoy_v3_api_field_config.listener.v3.Listener.name>`. By default static
+  // listeners cannot be modified or removed by LDS.
+  //
+  // Until LDS delivers a listener with a matching name the listener remains static, and an
+  // LDS state-of-the-world response that omits this name will not remove it. Once LDS does
+  // deliver a matching listener the static instance is replaced, and the replacement is
+  // thereafter treated as a regular dynamic listener (subject to LDS updates and removals).
```
Comment (AntonKanug) [inline api/envoy/config/listener/v3/listener.proto:458]: let me know if this behaviour makes sense, other option is give LDS SOTW response remove the listener if its not in the response.

**Comment by kyessenov** [comment]
Waiting for API review before code review.
/wait

**Comment by adisuissa** [comment]
Overriding a static resource (not necessarily a listener) with a dynamic one is discouraged due to the edge cases that it introduces.

Can you please add an explanation about the lifetime of the listeners.
Specifically, please describe cases where there's a statically defined listener, then a dynamic listener is added, and then it is removed. Please consider all the edge cases that you are aware of and add the relevant tests that cover them.

**Comment by wbpcode** [comment]
Thanks for this contribution. I am also be very curious about the actual scenarios that require this feature, could you add more background context about that? If dynamic updating is necessary, why not just use the dynamic listener directly? You can mark the Envoy instance be ready until the necessary listeners are ready?

**Comment by AntonKanug** [comment]
Our exact usecase is we run envoy as a sidecar and route a pod's DNS to envoy and use envoys DNS filter. We want the ability to override DNS resolutions to some known IPs. These resolution overrides can change at runtime so it needs to be set via the control plane.

As a result this DNS listener is dynamic, if envoy cant reach the control plane, pod's DNS fails entirely since its routed to a listener that doesnt exist. If we have a static config, we lose the resolution overrides but DNS doesnt fail entirely which is a better failure mode. 

One option is making that iptables routing to listener dynamic so that we can gate the routing on envoy having the listener setup and other is always having a static config and overriding it to a dynamic listener.

**Comment by AntonKanug** [comment]
@wbpcode does above make sense and thoughts on allowing handing off static listeners to be managed by LDS?

**Comment by wbpcode** [comment]
cc @AntonKanug I still think the dynamic `static` listener breaks Envoy's basic concept from API's perspective. cc @adisuissa

could we investigate the **ECDS** and FCDS to support the dynamic configuration of DNS gateway which could provide a default configuration? In my personal opinion, the ECDS should be the correct direction for your current requirements.

**Comment by AntonKanug** [comment]
Ah yeah ECDS covers the requirements thanks for that pointer, closing this out.
